devito/passes/iet/misc.py

```diff
@@ -77,47 +90,68 @@ def relax_incr_dimensions(iet, **kwargs):
 
         outer, inner = split(iterations, lambda i: not i.dim.parent.is_Incr)
 
-        # Compute the iteration ranges
-        ranges = []
-        for i in outer:
-            maxb = i.symbolic_max - (i.symbolic_size % i.dim.step)
-            ranges.append(((i.symbolic_min, maxb, i.dim.step),
-                           (maxb + 1, i.symbolic_max, i.symbolic_max - maxb)))
-
-        # Remove any offsets
-        # E.g., `x = x_m + 2 to x_M - 2` --> `x = x_m to x_M`
-        outer = [i._rebuild(limits=(i.dim.root.symbolic_min, i.dim.root.symbolic_max,
-                                    i.step))
-                 for i in outer]
-
-        # Create the ElementalFunction
-        name = sregistry.make_name(prefix="bf")
-        body = compose_nodes(outer)
-        dynamic_parameters = flatten((i.symbolic_bounds, i.step) for i in outer)
-        dynamic_parameters.extend([i.step for i in inner if not is_integer(i.step)])
-        efunc = make_efunc(name, body, dynamic_parameters)
-
-        efuncs.append(efunc)
-
-        # Create the ElementalCalls
-        calls = []
-        for p in product(*ranges):
-            dynamic_args_mapper = {}
-            for i, (m, M, b) in zip(outer, p):
-                dynamic_args_mapper[i.symbolic_min] = m
-                dynamic_args_mapper[i.symbolic_max] = M
-                dynamic_args_mapper[i.step] = b
-                for j in inner:
-                    if j.dim.root is i.dim.root and not is_integer(j.step):
-                        value = j.step if b is i.step else b
-                        dynamic_args_mapper[j.step] = (value,)
-            calls.append(efunc.make_call(dynamic_args_mapper))
-
-        mapper[root] = List(body=calls)
-
-    iet = Transformer(mapper).visit(iet)
-
-    return iet, {'efuncs': efuncs}
+        # Get symbolic_max out of each outer dimension
+        roots_max = {i.dim.root: i.symbolic_max for i in outer}
+
+        # A dictionary to map maximum of parent dimensions
+        # useful for hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+            assert i.direction is Forward
+            if (i.dim.parent in proc_parents_max and
+               i.symbolic_size == i.dim.parent.step):
+                # Special case: A parent dimension may have already been processed
+                # as a member of 'inner' iterations. In this case we can use parent's
+                # Max
+                # Usually encountered in hierarchical blocking (BLOCKLEVELS > 1)
+                it_max = proc_parents_max[i.dim.parent]
+            else:
+                # Most of the cases pass though this code:
+                # Candidates for upper bound calculation are:
```

Comment:
mathematically, there is a difference between "upper bound" and "max". Can you use the proper wording?

Reply:
Revisited